dartagnan/src/main/java/com/dat3m/dartagnan/parsers/program/visitors/spirv/VisitorSpirvOutput.java

```diff
@@ -0,0 +1,193 @@
+import com.dat3m.dartagnan.program.Program;
+import com.dat3m.dartagnan.program.memory.Location;
+import com.dat3m.dartagnan.program.memory.ScopedPointerVariable;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import static com.dat3m.dartagnan.expression.integers.IntCmpOp.*;
+import static com.dat3m.dartagnan.program.Program.SpecificationType.*;
+
+public class VisitorSpirvOutput extends SpirvBaseVisitor<Expression> {
+
+    private static final TypeFactory types = TypeFactory.getInstance();
+    private static final ExpressionFactory expressions = ExpressionFactory.getInstance();
+    private final Map<Location, Type> locationTypes = new HashMap<>();
+    private final ProgramBuilder builder;
+    private Program.SpecificationType type;
+    private Expression assertion;
+
+    public VisitorSpirvOutput(ProgramBuilder builder) {
+        this.builder = builder;
+    }
+
+    @Override
+    public Expression visitSpvHeaders(SpirvParser.SpvHeadersContext ctx) {
+        for (SpirvParser.SpvHeaderContext header : ctx.spvHeader()) {
+            if (header.outputHeader() != null && header.outputHeader().assertionList() != null) {
+                visitAssertionList(header.outputHeader().assertionList());
+            }
+        }
+        if (type == null) {
+            type = FORALL;
+            assertion = ExpressionFactory.getInstance().makeTrue();
+        }
+        builder.setSpecification(type, assertion);
+        return null;
+    }
+
+    @Override
+    public Expression visitAssertionList(SpirvParser.AssertionListContext ctx) {
+        Program.SpecificationType parsedType = parseType(ctx);
+        Expression parsedAssertion = ctx.assertion().accept(this);
+        appendAssertion(parsedType, parsedAssertion);
+        return null;
+    }
+
+    @Override
+    public Expression visitAssertionParenthesis(SpirvParser.AssertionParenthesisContext ctx) {
+        return ctx.assertion().accept(this);
+    }
+
+    @Override
+    public Expression visitAssertionNot(SpirvParser.AssertionNotContext ctx) {
+        return expressions.makeNot(ctx.assertion().accept(this));
+    }
+
+    @Override
+    public Expression visitAssertionAnd(SpirvParser.AssertionAndContext ctx) {
+        return expressions.makeAnd(ctx.assertion(0).accept(this), ctx.assertion(1).accept(this));
+    }
+
+    @Override
+    public Expression visitAssertionOr(SpirvParser.AssertionOrContext ctx) {
+        return expressions.makeOr(ctx.assertion(0).accept(this), ctx.assertion(1).accept(this));
+    }
+
+    @Override
+    public Expression visitAssertionBasic(SpirvParser.AssertionBasicContext ctx) {
+        Expression expr1 = ctx.assertionValue(0).accept(this);
+        Expression expr2 = ctx.assertionValue(1).accept(this);
+        IntCmpOp op = parseOperand(ctx.assertionCompare());
+        expr1 = normalize(expr1, expr2);
+        expr2 = normalize(expr2, expr1);
+        return expressions.makeBinary(expr1, op, expr2);
```

Comment:
The variables in assertions can have size different from the 64 bits of arch type. In this case it seems convenient to parse final constant values as 64 bit integers and then cast. Otherwise, we would need first to check if one of the operands is a variable and parse it first.

At some point we may also want to support assertions comparing different variable sizes (e.g., 32-bit and 64-bit integers) and cast to the size of the larger one.